File: leetcode/tree/1676_Lowest_Common_Ancestor_of_a_Binary_Tree_IV/codes/dfs.c
#include <stdbool.h>

#define MAX_NODES 10001

typedef struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
} TreeNode;

bool isTarget[MAX_NODES] = {0};

TreeNode* dfs(TreeNode* node) {
    if (!node) return NULL;
    if (isTarget[node -> val + 100000]) return node;
    TreeNode* left = dfs(node -> left);
    TreeNode* right = dfs(node -> right);
    if (left && right) return node;
    return left ? left : right;
}

TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode** nodes, int nodesSize) {
    for (int i = 0; i < nodesSize; i++) {
        isTarget[nodes[i] -> val + 100000] = true;
    }
    return dfs(root);
}

File: leetcode/tree/1676_Lowest_Common_Ancestor_of_a_Binary_Tree_IV/codes/dfs.py
from typing import List

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', nodes: 'List[TreeNode]') -> 'TreeNode':
        target_set = set(nodes)

        def dfs(node):
            if not node:
                return None
            if node in target_set:
                return node
            left = dfs(node.left)
            right = dfs(node.right)
            if left and right:
                return node
            return left if left else right

        return dfs(root)

        